### PR TITLE
Fix hex-based input boxes

### PR DESCRIPTION
**Pull request by X9VoiD**
## Changes

- Fixed normal hex parsing (was previously always zero, a regression from fixing hex byte arrays)
- Optimized `Allow` function
- Stricter hex input on both normal hex and hex byte array
- Removed duplicate code

### CODE REVIEW

File: OpenTabletDriver.UX/Windows/Configurations/Controls/DeviceIdentifierEditor.cs [hunk 1]
```diff
@@ -129,8 +129,8 @@ public DeviceIdentifierEntry()
 
             protected StackLayout layout;
 
-            private MaskedTextBox<int> vendorId, productId;
-            private MaskedTextBox<uint> inputReportLength, outputReportLength;
+            private HexNumberBox vendorId, productId;
+            private UnsignedIntegerNumberBox inputReportLength, outputReportLength;
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Windows/Configurations/Controls/DeviceIdentifierEditor.cs:133]: Why was this changed?

File: OpenTabletDriver.UX/Controls/Generic/Text/Providers/RegexBasedTextProvider.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+using System.Text.RegularExpressions;
+
+namespace OpenTabletDriver.UX.Controls.Generic.Text.Providers
+{
+    public abstract class RegexBasedTextProvider<T> : MaskedTextProvider<T>
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/Generic/Text/Providers/RegexBasedTextProvider.cs:5]: Would prefer this class name
```suggestion
    public abstract class RegexTextProvider<T> : MaskedTextProvider<T>
```

File: OpenTabletDriver.UX/Windows/Configurations/Controls/DeviceIdentifierEditor.cs (hunk 1 above)
Comment (X9VoiD) [reply OpenTabletDriver.UX/Windows/Configurations/Controls/DeviceIdentifierEditor.cs:133]: To make it clear what their type really are, instead of finding the `new()` just to know.

Comment (InfinityGhost) [reply OpenTabletDriver.UX/Windows/Configurations/Controls/DeviceIdentifierEditor.cs:133]: Reasonable enough

**Review by InfinityGhost — APPROVED** [review]
